drivers/dma/xilinx/qdma.c

```diff
@@ -0,0 +1,328 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * DMA driver for Xilinx QDMA Subsystem
+ *
+ * Copyright (C) 2022 Advanced Micro Devices, Inc. All rights reserved.
+ */
+
+#include <linux/mod_devicetable.h>
+#include <linux/regmap.h>
+#include <linux/dmaengine.h>
+#include <linux/platform_device.h>
+#include <linux/platform_data/amd_qdma.h>
+#include <linux/dma-mapping.h>
+#include "../virt-dma.h"
+#include "qdma-regs.h"
+
+/* mmio regmap config for all QDMA registers */
+static struct regmap_config qdma_regmap_config = {
+	.reg_bits = 32,
+	.val_bits = 32,
+	.reg_stride = 4,
+};
+
+/**
+ * struct qdma_request - DMA request structure
+ * @vdesc: Virtual DMA descriptor
+ * @dir: Transferring direction of the request
+ * @slave_addr: Physical address on DMA slave side
+ * @sgl: Scatter gather list on host side
+ * @sg_off: Start offset of the first sgl segment
+ * @nents: Number of sgl segments to transfer
+ */
+struct qdma_request {
+	struct virt_dma_desc		vdesc;
+	enum dma_transfer_direction	dir;
+	u64				slave_addr;
+	struct scatterlist		*sgl;
+	u32				sg_off;
+	u32				nents;
+};
+
+/**
+ * struct qdma_chan - Driver specific DMA channel structure
+ * @vchan: Virtual channel
+ * @qdev_hdl: Pointer to DMA device structure
+ * @busy: Busy flag of the channel
+ * @dir: Transferring direction of the channel
+ * @cfg: Transferring config of the channel
+ * @tasklet: Channel tasklet
+ */
+struct qdma_chan {
+	struct virt_dma_chan		vchan;
+	void				*qdev_hdl;
+	bool				busy;
+	enum dma_transfer_direction	dir;
+	struct dma_slave_config		cfg;
+	struct tasklet_struct		tasklet;
+};
+
+/**
+ * struct qdma_device - DMA device structure
+ */
+struct qdma_device {
+	struct platform_device 	*pdev;
+	struct dma_device	dma_dev;
+	struct regmap		*csr_regmap;
+	struct regmap		*intr_regmap;
+	struct regmap		*queue_regmap;
+};
+
+static inline struct qdma_chan *to_qdma_chan(struct dma_chan *chan)
+{
+	return container_of(chan, struct qdma_chan, vchan.chan);
+}
+
+static struct regmap *qdma_init_regmap(struct platform_device *pdev, char *name)
+{
+	struct regmap *regmap;
+	struct resource *res;
+	void *reg_base;
+
+	res = platform_get_resource_byname(pdev, IORESOURCE_MEM, name);
+	if (!res)
+		return NULL;
+
```

Comment:
res = platform_get_resource(pdev, IORESOURCE_IRQ, 0);

Reply:
Using by name is better than by index. The user driver does not need to know the index but using a friendly resource name. Please see latest kernel for platform_get_resource_byname()